### PR TITLE
Reinitiate the login flow using a cloned context

### PR DESCRIPTION
**Pull request by asha15**
### Proposed changes in this pull request

Currently session nonce cookie have been used to prevent the session hijack during the authentication. When a user switches the browser during the authentication (this mainly affects the mobile application where user switch from in app browser to the system browser) currently we’re displaying an error message, instead this PR propose to re-initiate the login flow. Here we're recreating the flow using a clone of the initial authentication context.

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java [hunk 1]
```diff
@@ -367,8 +368,11 @@ public void handle(HttpServletRequest request, HttpServletResponse response) thr
                 if (log.isDebugEnabled()) {
                     log.debug(e.getMessage(), e);
                 }
-                FrameworkUtils.sendToRetryPage(request, response, context, "suspicious.authentication.attempts",
-                        "suspicious.authentication.attempts.description");
+                log.warn("Session nonce cookie validation has failed. Hence, restarting the login flow.");
+                context = (AuthenticationContext) context.getProperty(FrameworkConstants.INITIAL_CONTEXT);
+                context.setProperty(FrameworkConstants.RESTART_LOGIN_FLOW, true);
+                FrameworkUtils.addAuthenticationContextToCache(context.getContextIdentifier(), context);
+                handle(request, response);
```
Comment (senthalan) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java:375]: Here we are adding a new entry to the Session table which will be redundant. 

can we add this `RESTART_LOGIN_FLOW` to the request and call handle method? Inside the handle method based on this request parameter, we can get back the initial context from the context and continue the flow.

Comment (asha15) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java]: Addressed with https://github.com/wso2/carbon-identity-framework/pull/4509/commits/b913f3378040a4c77e6f2376117ae1fa7f7f9009

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java [hunk 2]
```diff
@@ -367,16 +377,23 @@ public void handle(HttpServletRequest request, HttpServletResponse response) thr
                 if (log.isDebugEnabled()) {
                     log.debug(e.getMessage(), e);
                 }
-                FrameworkUtils.sendToRetryPage(request, response, context, "suspicious.authentication.attempts",
-                        "suspicious.authentication.attempts.description");
+                log.warn("Session nonce cookie validation has failed. Hence, restarting the login flow.");
```
Comment (asha15) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java:380]: add the sessionDataKey

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java [hunk 3]
```diff
@@ -184,6 +184,11 @@ public void handle(HttpServletRequest request, HttpServletResponse response,
                         }
                     } else if (context.getProperty(nonceCookieName) == null) {
                         addOrUpdateNonceCookie = true;
+                        // remove restartLoginFlow flag once we added session nonce cookie in the new browser session
```
Comment (senthalan) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultAuthenticationRequestHandler.java:187]:
```suggestion
                        // Remove restartLoginFlow flag once we added session nonce cookie in the new browser session.
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java [hunk 4]
```diff
@@ -206,10 +207,29 @@ public void handle(HttpServletRequest request, HttpServletResponse response) thr
                     request = FrameworkUtils.getCommonAuthReqWithParams(request, authRequest);
                 }
                 context = initializeFlow(request, responseWrapper);
+
+                //set the authentication request before the context is cloned.
```
Comment (senthalan) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java:211]:
```suggestion
                // Set the authentication request before the context is cloned.
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java [hunk 5]
```diff
@@ -206,10 +207,29 @@ public void handle(HttpServletRequest request, HttpServletResponse response) thr
...
+                if (authRequest != null) {
+                    context.setAuthenticationRequest(authRequest.getAuthenticationRequest());
+                }
+
+                // we'll use the cloned context to re-initiate the login flow when session
```
Comment (senthalan) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java:216]:
```suggestion
                // We'll use the cloned context to re-initiate the login flow when the session
```

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/servlet/CommonAuthenticationServlet.java [hunk 6]
```diff
@@ -44,17 +44,16 @@ public void init() {
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response)
             throws ServletException, IOException {
-        doPost(request, response);
+        CommonAuthenticationHandler commonAuthenticationHandler = new CommonAuthenticationHandler();
```
Comment (senthalan) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/servlet/CommonAuthenticationServlet.java:47]: let's make this `commonAuthenticationHandler` as class level variable

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/servlet/CommonAuthenticationServlet.java [hunk 7]
```diff
@@ -44,17 +44,16 @@ public void init() {
...
 
     @Override
     protected void doPost(HttpServletRequest request, HttpServletResponse response)
             throws ServletException, IOException {
 
-        if (FrameworkUtils.getMaxInactiveInterval() == 0) {
-            FrameworkUtils.setMaxInactiveInterval(request.getSession().getMaxInactiveInterval());
-        }
-        FrameworkUtils.getRequestCoordinator().handle(request, response);
+        CommonAuthenticationHandler commonAuthenticationHandler = new CommonAuthenticationHandler();
```
Comment (senthalan) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/servlet/CommonAuthenticationServlet.java:55]: let's make this commonAuthenticationHandler as class level variable

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/exception/CookieValidationFailedException.java [hunk 8]
```diff
@@ -0,0 +1,12 @@
+package org.wso2.carbon.identity.application.authentication.framework.exception;
+
+import java.io.IOException;
+
+/**
+ * Session nonce Cookie Validation Failed Exception.
+ */
+public class CookieValidationFailedException extends IOException {
+    public CookieValidationFailedException(String errorCode, String message) {
```
Comment (senthalan) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/exception/CookieValidationFailedException.java:9]: new line after the class definition

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/exception/CookieValidationFailedException.java [hunk 9]
```diff
@@ -0,0 +1,12 @@
+package org.wso2.carbon.identity.application.authentication.framework.exception;
```
[the remark this answers is not in the document]
Comment (asha15) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/exception/CookieValidationFailedException.java:19]: Addressed with https://github.com/wso2/carbon-identity-framework/pull/4509/commits/fbb5f9065181ab04c773a345028bfe9ebdba916f

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/servlet/CommonAuthenticationServlet.java (hunk 7 above)
Comment (asha15) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/servlet/CommonAuthenticationServlet.java]: Fixed in https://github.com/wso2/carbon-identity-framework/pull/4509/commits/fbb5f9065181ab04c773a345028bfe9ebdba916f

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/exception/CookieValidationFailedException.java (hunk 8 above)
Comment (asha15) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/exception/CookieValidationFailedException.java:27]: Fixed in https://github.com/wso2/carbon-identity-framework/pull/4509/commits/fbb5f9065181ab04c773a345028bfe9ebdba916f

**Review by senthalan — APPROVED** [review]

**Review by senthalan — APPROVED** [review]

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java [hunk 10]
```diff
@@ -206,10 +207,29 @@ public void handle(HttpServletRequest request, HttpServletResponse response) thr
...
+                }
+
                 context.initializeAnalyticsData();
+
             } else {
-                returning = true;
                 context = FrameworkUtils.getContextData(request);
+                if (request.getAttribute(FrameworkConstants.RESTART_LOGIN_FLOW) != null &&
+                        request.getAttribute(FrameworkConstants.RESTART_LOGIN_FLOW).equals("true")) {
+                    context = (AuthenticationContext) context.getProperty(FrameworkConstants.INITIAL_CONTEXT);
```
Comment (indeewari) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java:228]: This is not null safe. Earlier it was null safe. Please check.

**Pull request merged**